weather-station-oracle/smart-contract/contracts/WeatherStationOracle.sol

```diff
@@ -0,0 +1,77 @@
+pragma solidity ^0.5.8;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "iexec-doracle-base/contracts/IexecDoracle.sol";
+
+contract WeatherStationOracle is Ownable, IexecDoracle
+{
+	struct WeatherReport
+	{
+		bytes32 oracleCallID;
+		string 	description;
+		bool	historicalData;
+	}
+
+	struct Location
+	{
+		string  location;
+		mapping(uint256 => WeatherReport) reports;
+	}
+
+	mapping (bytes32 => Location) public locations;
+	
+
+	event NewWeatherReport(
+		bytes32 indexed id,
+		bytes32 indexed oracleCallID,
+		string 	location,
+		uint256 date,
+		string 	description
+	);
+
+	// Use _iexecHubAddr to force use of custom iexechub, leave 0x0 for autodetect
+	constructor(address _iexecHubAddr)
+	public IexecDoracle(_iexecHubAddr)
+	{}
+
+	function updateEnv(
+	  address _authorizedApp
+	, address _authorizedDataset
+	, address _authorizedWorkerpool
+	, bytes32 _requiredtag
+	, uint256 _requiredtrust
+	)
+	public onlyOwner
+	{
+		_iexecDoracleUpdateSettings(_authorizedApp, _authorizedDataset, _authorizedWorkerpool, _requiredtag, _requiredtrust);
+	}
+
+	function decodeResults(bytes memory results)
+	public pure returns(uint256, string memory, string memory)
+	{ return abi.decode(results, (uint256, string, string)); }
+
+	function processResult(bytes32 _oracleCallID)
+	public
+	{
+		uint256       date;
+		string memory location;
+		string memory description;
+
+		// Parse results
+		(date, location, description) = decodeResults(_iexecDoracleGetVerifiedResult(_oracleCallID));
+
+		// Process results
+		bytes32 id = keccak256(bytes(location));
+		require (!locations[id].reports[date].historicalData, "report is already fetched for given location and time");
+		emit NewWeatherReport(id, _oracleCallID, location, date, description);
+		locations[id].reports[date].oracleCallID 	= _oracleCallID;
+		locations[id].reports[date].description     = description;
+		locations[id].reports[date].historicalData 	= now > date ? true : false;
```

Comment:
This will always be false, as the time needed to push the report to the blockchain will not be instantaneous. I would remove it and replace line 66 with
`require(locations[id].reports[date].oracleCallID == bytes32(0))`

Reply:
It is possible to fetch forecast reports of the upcoming weather as future and past timestamps are valid. In the case of future weather reports, as the forecast can be updated with new data, it is possible to fetch a specific weather report again until the timestamp is in the past which is when historicalData will be true. And as we consider historical reports not to be updated, it is then not possible to fetch it again.